instrumentation/spring/spring-boot-resources/library/src/main/java/io/opentelemetry/instrumentation/spring/resources/SpringBootServiceNameGuesser.java

```diff
@@ -0,0 +1,278 @@
+        if (app != null) {
+          Object name = app.get("name");
+          return (String) name;
+        }
+      }
+    } catch (RuntimeException e) {
+      // expected to fail sometimes
+    }
+    return null;
+  }
+
+  @Nullable
+  private String findByCommandlineArgument() {
+    String result = attemptProcessHandleReflection();
+    if (result == null) {
+      String javaCommand = system.getProperty("sun.java.command");
+      result = parseNameFromCommandLine(javaCommand);
+    }
+    logger.log(Level.FINER, "Checking application commandline args: " + result);
+    return result;
+  }
+
+  @Nullable
+  private String attemptProcessHandleReflection() {
+    try {
+      String[] args = system.attemptGetCommandLineArgsViaReflection();
+      return parseNameFromProcessArgs(args);
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  @Nullable
+  private static String parseNameFromCommandLine(@Nullable String commandLine) {
+    if (commandLine == null) {
+      return null;
+    }
+    Matcher matcher = COMMANDLINE_PATTERN.matcher(commandLine);
+    if (matcher.find()) { // Required before group()
+      return matcher.group(1);
+    }
+    return null;
+  }
+
+  @Nullable
+  private static String parseNameFromProcessArgs(String[] args) {
+    return Stream.of(args)
+        .filter(arg -> arg.startsWith(COMMANDLINE_ARG_PREFIX))
+        .map(arg -> arg.substring(COMMANDLINE_ARG_PREFIX.length()))
+        .findFirst()
+        .orElse(null);
+  }
+
+  @Nullable
+  private String readNameFromAppProperties() {
+    return loadFromClasspath(
+        "application.properties", SpringBootServiceNameGuesser::getAppNamePropertyFromStream);
+  }
+
+  @Nullable
+  private static String getAppNamePropertyFromStream(InputStream in) {
+    Properties properties = new Properties();
+    try {
+      // Note: load() uses ISO 8859-1 encoding, same as spring uses by default for property files
+      properties.load(in);
+      return properties.getProperty("spring.application.name");
+    } catch (IOException e) {
+      return null;
+    }
+  }
+
+  @Nullable
+  private String loadFromClasspath(String filename, Function<InputStream, String> parser) {
+    try (InputStream in = system.openClasspathResource(filename)) {
+      return parser.apply(in);
+    } catch (Exception e) {
+      return null;
+    }
+  }
+
+  // Exists for testing
+  static class SystemHelper {
+
+    String getenv(String name) {
+      return System.getenv(name);
+    }
+
+    String getProperty(String key) {
+      return System.getProperty(key);
+    }
+
+    InputStream openClasspathResource(String filename) {
+      return ClassLoader.getSystemClassLoader().getResourceAsStream(filename);
+    }
+
+    InputStream openFile(String filename) throws Exception {
+      return Files.newInputStream(Paths.get(filename));
+    }
+
+    /**
+     * Attempts to use ProcessHandle to get the full commandline of the current process. Will only
+     * succeed on java 9+.
```

Comment:
can we use https://docs.oracle.com/javase/8/docs/api/java/lang/management/RuntimeMXBean.html#getInputArguments-- instead, to get Java 8 support?

Reply:
nm, I see this is for getting program args, maybe rename method to `attemptGetProgramArgsViaReflection`?

Reply:
Done 👍